### PR TITLE
ci: enable thin LTO on Mac

### PR DESCRIPTION
**Pull request by VerteDinde**
#### Description of Change

CircleCI has recently upped our base image size, raising the image from 84 GB to ~150 GB.

This PR re-enables thin LTO on Mac, and deletes the large thin LTO cache.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: none

### CODE REVIEW

File: .circleci/build_config.yml [hunk 1]
```diff
@@ -782,10 +690,6 @@ step-mksnapshot-build: &step-mksnapshot-build
     name: mksnapshot build
     command: |
       cd src
-      if [ "$USE_PREBUILT_V8_CONTEXT_SNAPSHOT" != "1" ]; then
-        ninja -C out/Default electron:electron_mksnapshot -j $NUMBER_OF_NINJA_PROCESSES
-        gn desc out/Default v8:run_mksnapshot_default args > out/Default/mksnapshot_args
-      fi
```
Comment (jkleinsc) [inline .circleci/build_config.yml:788]: This is still needed; just remove the if condition.

File: .circleci/build_config.yml [hunk 2]
```diff
@@ -798,10 +702,6 @@ step-mksnapshot-build: &step-mksnapshot-build
           electron/script/strip-binaries.py --file $PWD/out/Default/v8_context_snapshot_generator
         fi
       fi
-      if [ "$USE_PREBUILT_V8_CONTEXT_SNAPSHOT" != "1" ] && [ "$SKIP_DIST_ZIP" != "1" ]; then
-        ninja -C out/Default electron:electron_mksnapshot_zip -j $NUMBER_OF_NINJA_PROCESSES
-        (cd out/Default; zip mksnapshot.zip mksnapshot_args gen/v8/embedded.S)
-      fi
```
Comment (jkleinsc) [inline .circleci/build_config.yml:804]: This is still needed; just remove the if condition.

**Review by jkleinsc — CHANGES_REQUESTED** [review]
mksnapshot.zip is missing because some of the steps removed are still needed, they just no longer need the `if [ "$USE_PREBUILT_V8_CONTEXT_SNAPSHOT" != "1" ]; then` conditional as they should always run.

**Review by nornagon — APPROVED** [review]
excited to run out of 150 gb now :D

**Review by jkleinsc — APPROVED** [review]

**Comment by VerteDinde** [comment]
Looks like we run out of disk space on Mac arm64 publish steps if all of the original logic was removed 😭 Re-enabling USE_PREBUILT_V8_CONTEXT_SNAPSHOT to see if that leaves us with enough space.

File: .circleci/build_config.yml [hunk 3]
```diff
@@ -544,6 +538,8 @@ step-electron-build: &step-electron-build
         (cd out/Default; zip mksnapshot.zip mksnapshot_args clang_x64_v8_arm64/gen/v8/embedded.S)
         rm -rf out/Default/clang_x64_v8_arm64/gen
         rm -rf out/Default/clang_x64_v8_arm64/obj
+        rm -rf out/Default/clang_x64_v8_arm64/thinlto-cache
+        rm -rf out/Default/clang_x64_v8_arm64/v8_context_snapshot_generator.dSYM
```
Comment (MarshallOfSound) [inline .circleci/build_config.yml:542]: We still need this

Comment (VerteDinde) [reply .circleci/build_config.yml:542]: @MarshallOfSound Just the .dSYM file, right (the LTO cache is still fine to blow away)? I'll remove this, thanks for catching 👍

**Review by jkleinsc — CHANGES_REQUESTED** [review]
thin LTO is supposed to [decrease link time](https://source.chromium.org/chromium/chromium/src/+/main:build/config/compiler/compiler.gni;l=69;bpv=1;bpt=0), but it looks like enabling this on macOS makes our builds run longer, not shorter.  Compare https://app.circleci.com/pipelines/github/electron/electron/50376/workflows/fd850f0d-a752-457c-a1a5-0738e77173a4 to https://app.circleci.com/pipelines/github/electron/electron/50375/workflows/0c89f5a0-4d79-44cb-bc9e-23b0120627f6

**Review by jkleinsc — APPROVED** [review]
@VerteDinde pointed out that ThinLTO [achieves better runtime performance](https://clang.llvm.org/docs/ThinLTO.html), so while it means our builds will be slower, the end result is worth it.

**Pull request merged**